packages/lib/gpcircuits/src/list-membership.ts

```diff
@@ -0,0 +1,131 @@
+import { poseidon1, poseidon2, poseidon3, poseidon4} from "poseidon-lite";
+
+const poseidon = [poseidon2, poseidon3, poseidon4];
+
+export type ListMembershipModuleInputs = {
+  valueHash: CircuitSignal;
+  list: CircuitSignal[];
+};
+
+export type ListMembershipModuleInputNamesType = [
+  "valueHash",
+  "list"
+];
+
+export type ListMembershipModuleOutputs = { isMember: CircuitSignal };
+
+export type ListMembershipModuleOutputNamesType = ["isMember"];
+
+export class ListMembership {
+  // TODO: Rework this to allow for other modules that might make use of
+  // tuples.
+  /**
+   * Given indices of `N <= paramMaxValues` POD entry values and N lists
+   * of bigints, this procedure computes the signals necessary to carry out
+   * list membership checks (via `ListMembershipModule`) of each of the `N`
+   * value hashes in the corresponding list by forming appropriate lower
+   * order tuples and one combined membership list.
+   *
+   * @param `listIndices` denotes the indices of the POD entry values to be checked
+   * @param `membershipLists` is a list each of whose elements is a membership
+   *  list for the value associated with the corresponding index in `listIndices`.
+   * @param `paramMaxValues` denotes the maximum number of POD entry values in the
+   * circuit.
+   * @param `paramTupleArity` denotes the fixed tuple arity for the circuit.
+   * @param `paramMaxListEntries` denotes the maximum number of list entries in the circuit.
+   * @param `paramMaxTuples` denotes the maximum number of tuples in the circuit.
+   * @returns `number` containing the index of the hash whose membership
+   *  is to be checked, where an index less than `paramMaxValues` corresponds to
+   *  an entry value hash and a greater index corresponds to a tuple index.
+   * @returns `number[][]` containing `paramMaxTuples` lists of `paramTupleArity` tuples
+   *  of indices, where an index `i` refers to an entry value hash if it is less than
+   *  `paramMaxValues`, otherwise it refers to the tuple represented by the
+   *  `paramMaxValues - i`th entry of this list.
+   * @returns `bigint[]` containing the combined membership list.
+   * @throws TypeError if the lengths of `listIndices` and `membershipLists` don't agree.
+   * @throws TypeError if the list of all combinations of elements of `membershipLists`
+   *  exceeds `paramMaxListentries`.
+   * @throws TypeError if the tuple arity does not lie in the acceptable range.
+   * @throws TypeError if any element of `listIndices` exceeds `paramMaxValues - 1`.
+   * @throws TypeError if the number of tuples of arity `paramTupleArity` required to
+   *  represent `listIndices` exceeds `paramMaxTuples`.
+   */
+  public static generateSignals(listIndices: number[],
+				membershipLists: PODValue[][],
+				paramMaxValues: number,
+				paramTupleArity: number,
+				paramMaxListEntries: number,
+				paramMaxTuples: number
+			       ): [number, bigint[][], bigint[]] {
```

Comment:
clarity nit: Giving multiple return values names (by putting them in an object) can help make them self-documenting.